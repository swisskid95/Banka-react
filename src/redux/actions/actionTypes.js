export const LOGIN_USER = 'LOGIN_USER';
export const LOGIN_USER_ERROR = 'LOGIN_USER_ERROR';
export const SIGN_UP_USER = 'SIGN_UP_USER';
export const SIGN_UP_USER_ERROR = 'SIGN_UP_USER_ERROR';
export const ADD_ACCOUNT_DETAILS = 'ADD_ACCOUNT_DETAILS';
export const ADD_ALL_ACCOUNTS = 'ADD_ALL_ACCOUNTS';
export const ADD_ALL_ACCOUNTS_ERROR = 'ADD_ALL_ACCOUNTS_ERROR';
export const SET_CURRENT_USER = 'SET_CURRENT_USER';
export const GET_USER_DETAILS = 'GET_USER_DETAILS';
