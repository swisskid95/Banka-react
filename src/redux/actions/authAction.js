import { SET_CURRENT_USER } from './actionTypes';

export const setCurrentUser = decoded => {
  return {
    type: SET_CURRENT_USER,
    payload: decoded
  };
};

export const logout = () => {
  // remove token
  /* istanbul ignore next */
  localStorage.removeItem('user');
  /* istanbul ignore next */
  setCurrentUser({});
};
