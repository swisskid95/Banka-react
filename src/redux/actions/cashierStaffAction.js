import * as types from './actionTypes';
import axios from '../../utils/axiosConfig';
import { toast } from 'react-toastify';

export const getCustomerDetails = user => {
  return { type: types.GET_USER_DETAILS, payload: user };
};

export const fetchAccountDetails = (token, accountNumber) => {
  return async dispatch => {
    try {
      const response = await axios.get(`/accounts/${accountNumber}`, {
        headers: { Authorization: `Bearer ${token}` }
      });

      if (response.status === 200) {
        const data = response.data.data;

        dispatch(getCustomerDetails(data));
      }
    } catch (error) {
      /* istanbul ignore next */
      if (error.message === 'Network Error') {
        return toast.error('Something went wrong. Please try Again');
      }
      /* istanbul ignore next */
      const message = error.response.data.error;
      /* istanbul ignore next */
      toast.error(message);
    }
  };
};

export const performDeposit = (token, accountNumber, depositData) => {
  return async () => {
    try {
      const response = await axios.post(
        `/accounts/${accountNumber}/credit`,
        depositData,
        {
          headers: { Authorization: `Bearer ${token}` }
        }
      );

      if (response.status === 201) {
        const data = response.data.data;
        const message = `Successfully credited account with ${data.amount}`;
        toast.success(message);
      }
    } catch (error) {
      /* istanbul ignore next */
      if (error.message === 'Network Error') {
        return toast.error('Something went wrong. Please try Again');
      }
      /* istanbul ignore next */
      let message;
      /* istanbul ignore next */
      if (typeof error.response.data.error === 'object') {
        Object.values(error.response.data.error).join(', ');
      } else {
        /* istanbul ignore next */
        message = error.response.data.error;
      }
      /* istanbul ignore next */
      toast.error(message);
    }
  };
};
export const performWithdrawal = (token, accountNumber, depositData) => {
  return async () => {
    try {
      const response = await axios.post(
        `/accounts/${accountNumber}/debit`,
        depositData,
        {
          headers: { Authorization: `Bearer ${token}` }
        }
      );

      if (response.status === 201) {
        const data = response.data.data;
        const message = `Successfully debited account with ${data.amount}`;
        toast.success(message);
      }
    } catch (error) {
      /* istanbul ignore next */
      if (error.message === 'Network Error') {
        return toast.error('Something went wrong. Please try Again');
      }
      /* istanbul ignore next */
      let message;
      /* istanbul ignore next */
      if (typeof error.response.data.error === 'object') {
        Object.values(error.response.data.error).join(', ');
      } else {
        /* istanbul ignore next */
        message = error.response.data.error;
      }
      /* istanbul ignore next */
      toast.error(message);
    }
  };
};
