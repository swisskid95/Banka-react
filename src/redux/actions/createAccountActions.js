import * as types from './actionTypes';
import axios from '../../utils/axiosConfig';
import { toast } from 'react-toastify';

export const addAccountDetails = user => {
  return { type: types.ADD_ACCOUNT_DETAILS, payload: user };
};

export const createAccountRequest = (token, data) => {
  return async dispatch => {
    try {
      const response = await axios.post('/accounts', data, {
        headers: { Authorization: `Bearer ${token}` }
      });
      if (response.status === 201) {
        const data = response.data.data;
        const toastResponse = `A ${data.type} account with account number ${data.accountNumber} has been created`;
        toast.success(toastResponse);

        dispatch(addAccountDetails(data));
      }
    } catch (error) {
      /* istanbul ignore next */
      if (error.message === 'Network Error') {
        return toast.error('Something went wrong. Please try Again');
      }
      /* istanbul ignore next */
      const message = error.response.data.error;
      /* istanbul ignore next */
      toast.error(message);
    }
  };
};
