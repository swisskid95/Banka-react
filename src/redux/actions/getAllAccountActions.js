import * as types from './actionTypes';
import axios from '../../utils/axiosConfig';
import { toast } from 'react-toastify';

export const addAllAccounts = accounts => {
  return { type: types.ADD_ALL_ACCOUNTS, payload: accounts };
};

export const fetchAllAccountRequest = (token, emailAddress) => {
  return async dispatch => {
    try {
      const response = await axios.get(`users/${emailAddress}/accounts`, {
        headers: { Authorization: `Bearer ${token}` }
      });
      if (response.status === 200) {
        const data = response.data.data;
        dispatch(addAllAccounts(data));
      }
    } catch (error) {
      /* istanbul ignore next */
      if (error.message === 'Network Error') {
        return toast.error('Something went wrong. Please try Again');
      }
    }
  };
};
