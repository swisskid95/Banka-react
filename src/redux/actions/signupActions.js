import * as types from './actionTypes';
import axios from '../../utils/axiosConfig';
import { toast } from 'react-toastify';

export const signUpUser = user => {
  return { type: types.SIGN_UP_USER, payload: user };
};

export const signUpUserError = error => {
  return { type: types.SIGN_UP_USER_ERROR, payload: error };
};

export const signUpUserRequest = (user, history) => {
  return async dispatch => {
    try {
      const response = await axios.post('auth/signup', user);
      if (response.status === 201) {
        const data = response.data.data;
        localStorage.setItem('user', JSON.stringify(data));
        toast.success('Sign up Successful');
        setTimeout(() => {
          /* istanbul ignore next */
          history.push('/summary');
        }, 3000);
        dispatch(signUpUser(data));
      }
    } catch (error) {
      /* istanbul ignore next */
      if (error.message === 'Network Error') {
        return toast.error('Something went wrong. Please try Again');
      }
      /* istanbul ignore next */
      const message = error.response.data.error;
      /* istanbul ignore next */
      toast.error(message);
      /* istanbul ignore next */
      dispatch(signUpUserError(message));
    }
  };
};
