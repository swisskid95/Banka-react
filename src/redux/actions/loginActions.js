import * as types from './actionTypes';
import axios from '../../utils/axiosConfig';
import { toast } from 'react-toastify';
import { setCurrentUser } from './authAction';

export const logInUser = user => {
  return { type: types.LOGIN_USER, payload: user };
};

export const logInUserError = error => {
  return { type: types.LOGIN_USER_ERROR, payload: error };
};

export const loginUserRequest = (user, history) => {
  return async dispatch => {
    try {
      const response = await axios.post('auth/signin', user);
      if (response.status === 201) {
        const data = response.data.data;
        localStorage.setItem('user', JSON.stringify(data));
        toast.success('Log In Successful');
        dispatch(setCurrentUser(data));
        setTimeout(() => {
          /* istanbul ignore next */
          history.push('/summary');
        }, 3000);
        dispatch(logInUser(data));
      }
    } catch (error) {
      /* istanbul ignore next */
      if (error.message === 'Network Error') {
        return toast.error('Something went wrong. Please try Again');
      }
      /* istanbul ignore next */
      const message = error.response.data.error;
      /* istanbul ignore next */
      toast.error(message);
      /* istanbul ignore next */
      dispatch(logInUserError(message));
    }
  };
};
