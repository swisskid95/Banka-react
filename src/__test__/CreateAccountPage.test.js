/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable no-undef */
import {
  CreateAccountPage,
  mapDispatchToProps
} from '../components/CreateAccountPage';
import { ADD_ACCOUNT_DETAILS } from '../redux/actions/actionTypes';
import { createAccountRequest } from '../redux/actions/createAccountActions';
import { BrowserRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import mockAxios from 'axios';
import configureMockStore from 'redux-mock-store';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const handleSubmit = jest.fn();

const credential = {
  type: 'savings'
};

const user = {
  token:
    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MjMsI…c4OX0.DfJClt28K3qIgJJWp9FMypDO0tKD1NoiNtO9YvEOvsQ'
};
const renderCreateAccountPage = args => {
  const defaultProps = {
    createAccountRequest: jest.fn()
  };
  const props = { ...defaultProps, ...args, onSubmit: { handleSubmit } };
  return mount(
    <BrowserRouter>
      <CreateAccountPage {...props} />
    </BrowserRouter>
  );
};

describe('Create account in Components Tests', () => {
  it('renders create account page In template', () => {
    const wrapper = renderCreateAccountPage();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('form').length).toBe(1);
  });

  it(`Simulates an onChange event on select account type input`, () => {
    const wrapper = renderCreateAccountPage();
    wrapper
      .find('select')
      .at(0)
      .simulate('change', { currentTarget: { value: 'savings' } });
  });

  it(`Simulates an onchange event on form password input`, () => {
    const wrapper = renderCreateAccountPage();
    wrapper
      .find('input')
      .at(0)
      .simulate('change', { currentTarget: { value: 'Sanusi' } });
  });
  it('Simulates a form submit event', () => {
    const wrapper = renderCreateAccountPage();
    wrapper.find('form').simulate('submit');
  });

  it('should dispatch signup request action', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).createAccountRequest(user.token, credential);
  });
});

describe('create account request Async action Tests', () => {
  let store;
  beforeEach(() => {
    store = mockStore({});
    jest.resetAllMocks();
  });
  afterEach(() => {
    store.clearActions();
  });
  it('Should Trigger the ADD_ACCOUNT_DETAILS dispatch function', async () => {
    const mockData = {
      data: {
        data: {
          accountNumber: 9093234532,
          type: 'savings'
        }
      }
    };
    mockAxios.post.mockResolvedValue({
      status: 201,
      data: mockData
    });

    const expectedActions = [
      { type: ADD_ACCOUNT_DETAILS, payload: mockData.data }
    ];
    await store.dispatch(createAccountRequest(credential, user.token));
    expect(store.getActions()).toEqual(expectedActions);
  });
  //   it('Should Trigger the LOGIN_USER_ERROR dispatch function', async () => {
  //     const mockData = {
  //       data: {
  //         data: {
  //           message: 'invalid email/password'
  //         }
  //       }
  //     };
  //     mockAxios.post.mockResolvedValue({
  //       response: mockData
  //     });
  //     const expectedActions = [
  //       { type: SIGN_UP_USER_ERROR, payload: mockData.data.data.message }
  //     ];
  //     const { message } = mockData.data.data;
  //     store.dispatch(signUpUserError(message));
  //     expect(store.getActions()).toEqual(expectedActions);
  //   });
});

// describe('signup Reducer Tests', () => {
//   it('Should return default state', () => {
//     const newState = signupReducer(undefined, {});
//     expect(newState).toEqual({});
//   });
//   it('Should return a new state if it receive a signup action type', () => {
//     const user = {
//       firstName: 'fola',
//       lastName: 'abass',
//       email: 'abass@gmail.com',
//       token:
//         'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MjMsI…c4OX0.DfJClt28K3qIgJJWp9FMypDO0tKD1NoiNtO9YvEOvsQ'
//     };
//     const newState = signupReducer(undefined, {
//       type: SIGN_UP_USER,
//       payload: user
//     });
//     expect(newState).toEqual({ user });
//   });

//   it('Should return a new state if it receives a Signup error action type', () => {
//     const error = {
//       message: 'invalid email/password'
//     };
//     const newState = signupReducer(undefined, {
//       type: SIGN_UP_USER_ERROR,
//       payload: error
//     });
//     expect(newState).toEqual({ user: error });
//   });
// });
