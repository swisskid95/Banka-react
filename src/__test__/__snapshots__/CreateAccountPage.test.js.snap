// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create account in Components Tests renders create account page In template 1`] = `
<main
  className="container"
>
  <aside
    className="sidebar sidebar--light"
  >
    <span
      className="sidebar__close"
      onClick={[Function]}
    >
      Ã—
    </span>
    <div
      className="avatar"
    >
      <img
        alt="user"
        className="avatar__img"
        src="https://res.cloudinary.com/swisskid95/image/upload/v1566613597/Banka-assets/defaultAvatar_roxuye.png"
      />
    </div>
    <h3
      className="user-name"
    >
      undefined undefined
    </h3>
    <ul
      className="nav-container"
    >
      <li>
        <a
          aria-current={null}
          className="nav"
          href="/summary"
          onClick={[Function]}
        >
          summary
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          className="nav"
          href="/create-account"
          onClick={[Function]}
        >
          new account
        </a>
      </li>
      <li>
        <a
          className="nav  logout"
          href="/login"
          onClick={[Function]}
        >
          sign-out
        </a>
      </li>
    </ul>
  </aside>
  <div
    className="main-wrapper"
  >
    <div
      className="top-section"
    >
      <img
        alt="Sidebar open button"
        className="sidebar__open"
        onClick={[Function]}
        src="https://res.cloudinary.com/swisskid95/image/upload/v1566575800/Banka-assets/menu_ldfqrk.svg"
      />
      <a
        className="logo"
        href="/summary"
        onClick={[Function]}
      >
        <h1
          className="logo"
        >
          Banka
        </h1>
      </a>
    </div>
    <div
      className="content-title"
    >
      <h1
        className="content-title__text"
      >
        create account
      </h1>
    </div>
    <form
      className="entry--closer entry"
      onSubmit={[Function]}
    >
      <select
        className="entry__input--account-type entry__input entry__input--bigger"
        name="type"
        onBlur={[Function]}
      >
        <option
          value="current"
        >
          current
        </option>
        <option
          value="savings"
        >
          savings
        </option>
      </select>
      <input
        className="entry__input entry__input--bigger"
        name="accountPurpose"
        onChange={[Function]}
        placeholder="account purpose"
        type="text"
      />
      <button
        className="btn entry__btn entry__input--bigger "
      >
        create account
      </button>
    </form>
  </div>
</main>
`;
